README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# shinyB2AIDashboard

<!-- badges: start -->
<!-- badges: end -->

The goal of shinyB2AIDashboard is to support dashboarding for the
Bridge2AI Teaming Core. It contains functionality to:

- Collate data about Bridge2AI Core and data generation program (DGP)
  members
- Collate publications for Program members
- Develop R shiny-based dashboard(s)

Still to do is to develop said dashboards

## Installation

You can install the development version of shinyB2AIDashboard like so:

``` r
library(remotes)
remotes::install_github('seandavi/shinyB2AIDashboard')
```

## Example

This is a basic example which shows you how to solve a common problem:

``` r
library(shinyB2AIDashboard)
```

### Fetch all publications by orcid

``` r
orcid_example = '0000-0002-8991-6458'
works_df = works_from_orcid(orcid_example)
```

The result is a data.frame with some minimal information about each
publication.

``` r
colnames(works_df)
#> [1] "title"    "doi"      "pmid"     "pmc"      "put_code" "orcid"
dim(works_df)
#> [1] 98  6
head(works_df)
#>                                                                                                                              title
#> 1 GenomicSuperSignature facilitates interpretation of RNA-seq experiments through robust, efficient comparison to public databases
#> 2       HGNChelper: identification and correction of invalid gene symbols for human and mouse [version 2; peer review: 3 approved]
#> 3                                                                                 Ten simple rules for large-scale data processing
#> 4            GenomicSuperSignature: interpretation of RNA-seq experiments through robust, efficient comparison to public databases
#> 5                                            HGNChelper: identification and correction of invalid gene symbols for human and mouse
#> 6                                            HGNChelper: identification and correction of invalid gene symbols for human and mouse
#>                              doi pmid  pmc  put_code               orcid
#> 1     10.1038/s41467-022-31411-3 <NA> <NA> 115044512 0000-0002-8991-6458
#> 2 10.12688/f1000research.28033.2 <NA> <NA>  85614478 0000-0002-8991-6458
#> 3   10.1371/journal.pcbi.1009757 <NA> <NA> 107995204 0000-0002-8991-6458
#> 4      10.1101/2021.05.26.445900 <NA> <NA>  94544767 0000-0002-8991-6458
#> 5 10.12688/f1000research.28033.1 <NA> <NA>  85615672 0000-0002-8991-6458
#> 6      10.1101/2020.09.16.300632 <NA> <NA>  83594630 0000-0002-8991-6458
```

### Fetch details about one of the publications

``` r
work_details = orcid_work_detail(orcid_example, works_df$put_code[6])
dplyr::glimpse(work_details)
#> Rows: 1
#> Columns: 13
#> $ orcid          <chr> "0000-0002-8991-6458"
#> $ put_code       <int> 83594630
#> $ created_at     <dttm> 2020-11-17 11:45:46
#> $ updated_at     <dttm> 2022-05-31 08:33:36
#> $ citation_type  <chr> "bibtex"
#> $ citation_value <chr> "@article{Oh_2020,\n\tdoi = {10.1101/2020.09.16.300632}…
#> $ work_type      <chr> "other"
#> $ contributors   <lgl> NA
#> $ url            <chr> "https://doi.org/10.1101/2020.09.16.300632"
#> $ doi            <chr> "10.1101/2020.09.16.300632"
#> $ pmid           <lgl> NA
#> $ pmc            <lgl> NA
#> $ abstract       <lgl> NA
```

## Dashboards

Work in progress….

## The Bridge2AI Program

The NIH Common Fund’s Bridge to Artificial Intelligence (Bridge2AI)
program will propel biomedical research forward by setting the stage for
widespread adoption of artificial intelligence (AI) that tackles complex
biomedical challenges beyond human intuition. A key step in this process
is generating new “flagship” data sets and best practices for machine
learning analysis. Machine learning is a type of Artificial Intelligence
(AI) that provides machines (like computers) the ability to
automatically learn and improve from experience, without being
explicitly programmed. The biomedical research community generates a
wealth of data, but most of these data are not suitable for machine
learning because they are incomplete. By bringing technological and
biomedical experts together with social scientists and humanists, the
Bridge2AI program will help bring solutions to this deficit by:

- Generating new flagship biomedical and behavioral data sets that are
  ethically sourced, trustworthy, well-defined, and accessible
- Developing software and standards to unify data attributes across
  multiple data sources and across data types
- Creating automated tools to accelerate the creation of FAIR (Findable,
  Accessible, Interoperable, and Reusable) and ethically sourced data
  sets
- Providing resources to disseminate data, ethical principles, tools,
  and best practices
- Creating training materials and activities for workforce development
  that bridges the AI, biomedical, and behavioral research communities


NEWS.md
# shinyB2AIDashboard 0.1.3

* Perform some cleanup based on devtools::check() output

# shinyB2AIDashboard 0.1.2

* Beautifying a bit with pkgdown settings

# shinyB2AIDashboard 0.1.1

* Update README.Rmd and README.md with more detail

# shinyB2AIDashboard 0.1.0

## New functionality

* Add function `dssrc_dataframe` to fetch b2ai program onboarding dataset
  from google spreadsheet
* Add function `orcid_api_call` as a simple client for the 
  [ORCID API](https://pub.orcid.org/v3.0/#/Development_Public_API_v3.0).
* Add function `works_from_orcid` to get all the publication 
  records for a given ORCID.
* Add function `orcid_work_detail` to fetch the details of a given
  ORCID work record. 
* A number of helper functions support the end-user functions above.

## Other

* Add documentation for new functions.
* Add _pkgdown.yml base.
* Add tests for `orcid_api_call`
* Add tests for `works_from_orcid`
* Basic pkgdown website published as gh-pages

# shinyB2AIDashboard 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(shinyB2AIDashboard)

test_check("shinyB2AIDashboard")


tests/testthat/test-orcid.R
library(testthat)

# this orcid has pmc, doi, and pmid values, so use for convenience
# many orcids will not have any pmid or pmc values
orcid_test = '0000-0001-5643-4068'
works_df = works_from_orcid(orcid_test)

test_that("works_from_orcid returns correct structure", {
  # get a data.frame
  expect_s3_class(works_df, 'data.frame')

  # six columns
  expect_equal(ncol(works_df),6)

  # check column names
  expect_identical(colnames(works_df),c('title','doi','pmid','pmc','put_code','orcid'))

  # at least one doi is not NA
  expect_gt(sum(!is.na(works_df$doi)),0)

  # at leasd one pmid is not NA
  expect_gt(sum(!is.na(works_df$pmid)),0)

  # at least one pmc is not NA
  expect_gt(sum(!is.na(works_df$pmc)),0)

  # no NA put_codes
  expect_equal(sum(is.na(works_df$put_code)),0)

  # double-check that return orcid matches input
  expect_equal(works_df$orcid[1], orcid_test)
})


tests/testthat/test-orcid_api.R
orcid_test = '0000-0001-5643-4068'

test_that("orcid_api returns correctly", {
  json_resp = orcid_api_call(orcid_test)
  expect_equal(class(json_resp),'list')
  expect_equal(names(json_resp), c("last-modified-date", "group", "path"))
})


R/utils.R
#' Test whether an object with names has specified key
#'
#' Test whether a list, data.frame(columns), or a named
#' vector has a specified key in it.
#'
#' @param l an object with names
#' @param key the name that you want to retrieve
#'
#'
#' @family utilities
#'
#' @keywords Internal
#'
#' @examples
#' l = list(a=1,b=2,c=3)
#' has_key(l, 'b') #TRUE
#' has_key(l, 'not_present') #FALSE
#'
#' @export
has_key <- function(l, key) {
    return(key %in% names(l))
}


#' Get a value from an object with names
#'
#' @param l an object with names
#' @param key the name that you want to retrieve
#' @param default the default value to return if the
#'    key does not exist
#'
#'
#' @family utilities
#'
#' @keywords Internal
#'
#' @examples
#' l = list(a=1,b=2,c=3)
#' safe_get(l,'a')
#' safe_get(l, 'not_present')
#'
#' @export
safe_get <- function(l, key, default=NA) {
    if(has_key(l, key)) {
        res = get(key, l)
        if(is.null(res)) {
            return(default)
        }
        return(get(key, l))
    }
    return(default)
}


R/orcid_api.R
#' Helper function for calling orcid API
#'
#' The ORCID REST API does not require any authentication for basic
#' querying. The OpenAPI documentation is available.
#'
#' The API typically works with the ORCID as the base resource. Therefore,
#' an ORCID is required for this function. The path follows the documentation.
#'
#' @param orcid character(1) the required orcid
#' @param works character(1) the rest of the API path after the `orcid` part.
#'
#' @returns an R `list` representation of the REST API json response
#'
#' @examples
#' orcid='0000-0001-7443-925X'
#'
#' orcid_api_call(orcid)
#'
#' @export
orcid_api_call <- function(orcid, path='/works') {
    url = sprintf('https://pub.orcid.org/v3.0/%s%s',orcid, path)

    # TODO: error checking
    resp = httr::GET(url, httr::accept_json())

    return(httr::content(resp))
}


R/orcid_work_resource.R
.extract_external_ids <- function(external_id_list) {
    external_id_recs = external_id_list$`external-id`
    retlist = list()
    for(i in seq_along(external_id_recs)) {
        id_type = external_id_recs[[i]]$`external-id-type`
        if(id_type %in% c('doi', 'pmid', 'pmc')) {
            value = external_id_recs[[i]]$`external-id-value`
            retlist[[id_type]] = ifelse(is.null(value), NA, value)
        }
    }
    return(retlist)
}

#' Extract a simple list from work-summary record
#'
#' This internal helper function extracts the title
#' and external ids (pmid, pmc, and doi) where available.
#'
#' @family orcid
#' @family bibliography
#'
#' @keywords Internal
.extract_simple_details_from_work_record <- function(work) {
    work_summary = work$`work-summary`[[1]]

    retlist = list(title=NA, doi=NA, pmid=NA, pmc = NA, put_code=NA)

    retlist['title'] = safe_get(work_summary$title$title,'value')
    retlist['put_code'] = safe_get(work_summary, 'put-code')

    # We're going to look for these `types` of
    # external records


    # this mess crawls through the external-id records
    # to extract the `type` and the `value`
    if(has_key(work_summary,'external-ids')) {
        external_id_recs = work_summary$`external-ids`
        external_ids = .extract_external_ids(external_id_recs)
        for(i in names(external_ids)) {
            retlist[[i]] = external_ids[[i]]
        }
    }

    retlist = data.frame(retlist)

    return(retlist)
}


#' Fetch raw JSON for ORCID works
#'
#'
#' @family orcid
#' @family bibliography
#'
#' @examples
#' json_response = orcid_works_json('0000-0003-3865-7844')
#'
#' str(json_response)
#'
#' @export
orcid_works_json <- function(orcid) {
    json_resp = orcid_api_call(orcid, path='/works')
    return(json_resp)
}

#' Works from a single ORCID
#'
#' Fetch a (simplified) set of works (publications, etc)
#' from an ORCID.
#'
#' @param orcid character(1) an orcid is required
#'
#' @returns a dataframe (see example)
#'
#' @examples
#' orcid = '0000-0001-5643-4068'
#'
#' df = works_from_orcid(orcid)
#' dim(df)
#' head(df)
#' summary(df)
#'
#' @family orcid
#' @family bibliography
#'
#' @export
works_from_orcid <- function(orcid) {
    json_resp = orcid_api_call(orcid,path='/works')
    works_list = json_resp$group
    if(length(works_list)==0) {
        return(data.frame(title=NULL,doi=NULL,pmid=NULL,pmc=NULL,put_code=NULL,orcid=NULL))
    }
    simple_works = lapply(works_list, .extract_simple_details_from_work_record)
    simple_works_df = data.frame(do.call(rbind, simple_works))
    simple_works_df$orcid = orcid
    return(simple_works_df)
}

#' Get a ORCID `work` detail
#'
#' The ORCID API has the concept of a `work` with its
#' own dedicated API endpoint. This function calls
#' that endpoint and returns a single data.frame
#' row representing that `work`.
#'
#' Additional functionality is available if the
#' `work` record has a bibtex entry. Using the `rbibutils`
#' package, one can convert the citation to bibtex, latex,
#' citations of various styles, and even HTML.
#'
#' @param orcid character(1) an orcid is required
#' @param put_code integer(1) the put-code, an ORCID API internal identifier
#'     for a work
#'
#' @family orcid
#' @family bibliography
#'
#' @examples
#' orcid = '0000-0001-5643-4068'
#' put_code = 89488980
#'
#' df = orcid_work_detail(orcid, put_code)
#' df
#' colnames(df)
#'
#' # the bibtex entry, if available,
#' # can be used for additional manipulation
#' if(require('rbibutils')) {
#'    rref = rbibutils::charToBib(df$citation_value)
#'    print(rref)
#'
#'    # bibtex
#'    print(rref, style = "Bibtex")
#'
#'    # simple citation
#'    print(rref, style = "citation")
#'
#'    # html
#'    print(rref, style = "html")
#'
#'    # latex
#'    print(rref, style = "latex")
#'
#'    # r citation
#'    print(rref, style = "R")
#' }
#'
#' @export
orcid_work_detail <- function(orcid, put_code) {
    retlist = list(
        orcid = orcid,
        put_code = put_code,
        created_at = NA,
        updated_at = NA,
        citation_type = NA,
        citation_value = NA,
        work_type = NA,
        contributors = NA,
        url = NA,
        doi = NA,
        pmid = NA,
        pmc = NA,
        abstract = NA
    )
    json_resp = orcid_api_call(orcid, path=sprintf('/work/%d', as.integer(put_code)))

    base = json_resp
    retlist[['created_at']] = lubridate::as_datetime(base$`created-date`$value/1000)
    retlist[['updated_at']] = lubridate::as_datetime(base$`last-modified-date`$value/1000)
    retlist[['citation_type']] = safe_get(base$citation, 'citation-type')
    retlist[['citation_value']] = safe_get(base$citation, 'citation-value')
    retlist[['url']] = safe_get(base$url,'value')
    retlist[['work_type']] = safe_get(base,'type')
    retlist[['abstract']] = safe_get(base,'short-description')

    if(has_key(base,'external-ids')) {
        external_id_recs = base$`external-ids`
        external_ids = .extract_external_ids(external_id_recs)
        for(i in names(external_ids)) {
            retlist[[i]] = external_ids[[i]]
        }
    }

    return(data.frame(retlist))
}

